File: examples/frameworks/python/langchain_agent.py
import os
import uuid

from langchain.agents import create_agent
from langchain_openai import ChatOpenAI
from rebuno import Agent, http_client, step
from support import TOOLS


async def process(query: str) -> dict:
    ref = await step("brief_ref", lambda: uuid.uuid4().hex[:8])
    llm = ChatOpenAI(
        model=os.environ["LLM_MODEL"],
        http_async_client=http_client(),
        streaming=os.environ.get("LLM_STREAM") == "1",
    )
    graph = create_agent(model=llm, tools=TOOLS)
    result = await graph.ainvoke(
        {"messages": [{"role": "user", "content": f"[{ref}] {query}"}]}
    )
    return {"answer": result["messages"][-1].content, "ref": ref}


agent = Agent(os.environ.get("AGENT_ID", "support-py-langchain"))

if __name__ == "__main__":
    agent.run(process, port=int(os.environ.get("AGENT_PORT", "5000")))

File: examples/frameworks/typescript/langchain_agent.ts
import { randomUUID } from "node:crypto";
import { env } from "node:process";

import { tool } from "@langchain/core/tools";
import { ChatOpenAI } from "@langchain/openai";
import { createAgent } from "langchain";
import { Agent, rebunoFetch, step } from "rebuno";

import { supportTools } from "./support.js";

async function process(input: { query: string }) {
  const modelName = env.LLM_MODEL;
  if (!modelName) throw new Error("LLM_MODEL is required");
  const ref = await step("brief_ref", () => randomUUID().replaceAll("-", "").slice(0, 8));
  const model = new ChatOpenAI({ model: modelName, configuration: { fetch: rebunoFetch } });

  // The agent loop keeps calling the model after a tool is held for approval,
  // and each call is refused. Aborting on the first one ends the loop instead.
  const halt = new AbortController();
  const tools = supportTools.map(({ name, description, schema, execute }) => tool(
    async (args: Record<string, unknown>) => {
      try { return await execute(args); }
      catch (error) { halt.abort(error); throw error; }
    },
    { name, description, schema },
  ));

  const graph = createAgent({ model, tools });
  const result = await graph.invoke(
    { messages: [{ role: "user", content: `[${ref}] ${input.query}` }] },
    { signal: halt.signal },
  );
  return { answer: result.messages.at(-1)?.content, ref };
}

const agent = new Agent(env.AGENT_ID ?? "support-ts-langchain");
await agent.serve({ port: Number(env.AGENT_PORT ?? 5000) }, process);
